In [1]:
import pandas as pd

In [4]:
pip install xlrd

You are using pip version 19.0.3, however version 20.2.3 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [7]:
df = pd.read_excel('default_of_credit_card_clients.xls')

In [8]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

In [10]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

In [11]:
df.ID.nunique()

29687

In [12]:
df.shape

(30000, 25)

We see that the number of unique ID's is less then the number of rows, hence, IDs are not entirely unique.

In [13]:
id_counts = df.ID.value_counts()
id_counts.head()

77222291-e191    2
91321f74-e4a4    2
20b16b21-8b23    2
4e2380e6-a8cf    2
96970f37-abaa    2
Name: ID, dtype: int64

In [14]:
#generate value counts for redundant ids
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

We see that there are 313 IDs that are repeated twice

In [15]:
import numpy as np

In [16]:
np.random.seed(seed=24)

In [18]:
random_integers = np.random.randint(low=1,high=5,size=100)
random_integers.shape

(100,)

In [20]:
random_integers[0:5]

array([1, 4, 3, 2, 3])

In [22]:
is_equal_to_3 = random_integers == 3

In [23]:
is_equal_to_3[:5]


array([False, False,  True, False,  True])

In [24]:
sum(is_equal_to_3)


27

In [25]:
random_integers[is_equal_to_3]

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3])

In [26]:
id_counts = df['ID'].value_counts()
dupe_mask = id_counts == 2
dupe_mask[0:5]

77222291-e191    True
91321f74-e4a4    True
20b16b21-8b23    True
4e2380e6-a8cf    True
96970f37-abaa    True
Name: ID, dtype: bool

In [27]:
dupe_ids = id_counts.index[dupe_mask]
dupe_ids = list(dupe_ids)
len(dupe_ids)

313

In [29]:
df.loc[df['ID'].isin(dupe_ids[0:3]),:].head(10) 

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
13186  20b16b21-8b23     750000    2          1         2   28     1     -1   
13286  20b16b21-8b23          0    0          0         0    0     0      0   
15190  91321f74-e4a4      30000    2          2         2   24     3      2   
15290  91321f74-e4a4          0    0          0         0    0     0      0   
24654  77222291-e191     160000    1          1         1   39    -2     -2   
24754  77222291-e191          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
13186     -1      0  ...     399493     400993     409483    250000    401003   
13286      0      0  ...          0          0          0         0         0   
15190      2      7  ...        300        300        300         0         0   
15290      0      0  ...          0          0          0         0         0   
24654     -2     -2  ...          0       2920          0        35         0   
24754      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
13186      7990     11000    417990      3080                           0  
13286         0         0         0         0                           0  
15190         0         0         0         0                           0  
15290         0         0         0         0                           0  
24654         0      2920         0     12140                           0  
24754         0         0         0         0                           0  

[6 rows x 25 columns]

In [34]:
df_zero_mask = df == 0
df_zero_mask.head()

ID  LIMIT_BAL    SEX  EDUCATION  MARRIAGE    AGE  PAY_1  PAY_2  PAY_3  \
0  False      False  False      False     False  False  False  False  False   
1  False      False  False      False     False  False  False  False   True   
2  False      False  False      False     False  False   True   True   True   
3  False      False  False      False     False  False   True   True   True   
4  False      False  False      False     False  False  False   True  False   

   PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  False  ...       True       True       True      True     False      True   
1   True  ...      False      False      False      True     False     False   
2   True  ...      False      False      False     False     False     False   
3   True  ...      False      False      False     False     False     False   
4   True  ...      False      False      False     False     False     False   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0      True      True      True                       False  
1     False      True     False                       False  
2     False     False     False                        True  
3     False     False     False                        True  
4     False     False     False                        True  

[5 rows x 25 columns]

In [35]:
feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)
sum(feature_zero_mask)

315

In [36]:
df_clean_1 = df.loc[~feature_zero_mask,:].copy()

In [37]:
df_clean_1['ID'].nunique()

29685

In [38]:
df_clean_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29685 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          29685 non-null  object
 1   LIMIT_BAL                   29685 non-null  int64 
 2   SEX                         29685 non-null  int64 
 3   EDUCATION                   29685 non-null  int64 
 4   MARRIAGE                    29685 non-null  int64 
 5   AGE                         29685 non-null  int64 
 6   PAY_1                       29685 non-null  object
 7   PAY_2                       29685 non-null  int64 
 8   PAY_3                       29685 non-null  int64 
 9   PAY_4                       29685 non-null  int64 
 10  PAY_5                       29685 non-null  int64 
 11  PAY_6                       29685 non-null  int64 
 12  BILL_AMT1                   29685 non-null  int64 
 13  BILL_AMT2                   29685 non-null  in

In [39]:
df_clean_1.PAY_1.head(15)

0      2
1     -1
2      0
3      0
4     -1
5      0
6      0
7      0
8      0
9     -2
10     0
11    -1
12    -1
13     1
14     0
Name: PAY_1, dtype: object

In [40]:
df_clean_1['PAY_1'].value_counts()

0                13087
-1                5047
1                 3261
Not available     3021
-2                2476
2                 2378
3                  292
4                   63
5                   23
8                   17
6                   11
7                    9
Name: PAY_1, dtype: int64

In [42]:
valid_pay_1_mask = df_clean_1['PAY_1'] != 'Not available'
valid_pay_1_mask[0:5]

0    True
1    True
2    True
3    True
4    True
Name: PAY_1, dtype: bool

In [43]:
sum(valid_pay_1_mask)

26664

In [45]:
df_clean_2 = df_clean_1.loc[valid_pay_1_mask,:].copy()
df_clean_2.shape

(26664, 25)

In [46]:
df_clean_2.PAY_1.value_counts()

 0    13087
-1     5047
 1     3261
-2     2476
 2     2378
 3      292
 4       63
 5       23
 8       17
 6       11
 7        9
Name: PAY_1, dtype: int64

In [47]:
df_clean_2['PAY_1'] = df_clean_2['PAY_1'].astype('int64')
df_clean_2[['PAY_1', 'PAY_2']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26664 entries, 0 to 29999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   PAY_1   26664 non-null  int64
 1   PAY_2   26664 non-null  int64
dtypes: int64(2)
memory usage: 624.9 KB


In [24]:
df_clean_2['EDUCATION'].replace(to_replace=[0, 5, 6], value=4, inplace=True)
df_clean_2['EDUCATION'].value_counts()

2    12458
1     9412
3     4380
4      414
Name: EDUCATION, dtype: int64

In [25]:
df_clean_2['MARRIAGE'].replace(to_replace=0, value=3, inplace=True)
df_clean_2['MARRIAGE'].value_counts()

2    14158
1    12172
3      334
Name: MARRIAGE, dtype: int64

In [26]:
df_clean_2['EDUCATION_CAT'] = 'none'


In [27]:
df_clean_2[['EDUCATION', 'EDUCATION_CAT']].head(10)


EDUCATION EDUCATION_CAT
0          2          none
1          2          none
2          2          none
3          2          none
4          2          none
5          1          none
6          1          none
7          2          none
8          3          none
9          3          none

In [28]:
cat_mapping = {
    1: "graduate school",
    2: "university",
    3: "high school",
    4: "others"
}

In [29]:
df_clean_2['EDUCATION_CAT'] = df_clean_2['EDUCATION'].map(cat_mapping)
df_clean_2[['EDUCATION', 'EDUCATION_CAT']].head(10)

EDUCATION    EDUCATION_CAT
0          2       university
1          2       university
2          2       university
3          2       university
4          2       university
5          1  graduate school
6          1  graduate school
7          2       university
8          3      high school
9          3      high school

In [32]:
edu_ohe = pd.get_dummies(df_clean_2['EDUCATION_CAT'])
edu_ohe.head(10)
df_with_ohe = pd.concat([df_clean_2, edu_ohe], axis=1)
df_with_ohe[['EDUCATION_CAT', 'graduate school',
             'high school', 'university', 'others']].head(10)
df_with_ohe.to_csv('cleaned_data.csv', index=False)
